File: decision-matrix/frontend/src/lib/padEarthworkSketch/envelope.ts
import { clampVertex, polygonPerimeterM } from './polygon';
import { localPlanCorners } from './rectangle';
import { isPlanPolygon, type PlanShapeSketch, type PlanVertex } from './types';

export type EnvelopeWrapParams = {
  enabled: boolean;
  wrap_width_m: number;
};

export const DEFAULT_ENVELOPE_WRAP_WIDTH_M = 3;

function edgeOutwardNormal(dx: number, dy: number): { x: number; y: number } {
  const len = Math.hypot(dx, dy);
  if (len < 1e-12) return { x: 0, y: 0 };
  return { x: dy / len, y: -dx / len };
}

/** Outward offset polygon (CCW vertices). Miter with bevel clamp on sharp angles. Negative distance = inward. */
export function offsetPolygonOutward(vertices: PlanVertex[], distance: number): PlanVertex[] {
  const n = vertices.length;
  if (n < 3 || distance === 0) return vertices.map((v) => ({ ...v }));
  const dist = Math.abs(distance);
  const sign = distance < 0 ? -1 : 1;
  const maxMiter = dist * 4;
  const out: PlanVertex[] = [];
  for (let i = 0; i < n; i += 1) {
    const prev = vertices[(i - 1 + n) % n];
    const curr = vertices[i];
    const next = vertices[(i + 1) % n];
    const e1x = curr.east_m - prev.east_m;
    const e1y = curr.north_m - prev.north_m;
    const e2x = next.east_m - curr.east_m;
    const e2y = next.north_m - curr.north_m;
    const n1 = edgeOutwardNormal(e1x, e1y);
    const n2 = edgeOutwardNormal(e2x, e2y);
    let bx = n1.x + n2.x;
    let by = n1.y + n2.y;
    const bl = Math.hypot(bx, by);
    let ox: number;
    let oy: number;
    if (bl < 1e-9) {
      ox = n1.x * dist;
      oy = n1.y * dist;
    } else {
      bx /= bl;
      by /= bl;
      const dot = n1.x * bx + n1.y * by;
      const scale = Math.abs(dot) > 1e-6 ? dist / dot : dist;
      if (!Number.isFinite(scale) || scale < 0 || scale > maxMiter) {
        ox = n1.x * dist;
        oy = n1.y * dist;
      } else {
        ox = bx * scale;
        oy = by * scale;
      }
    }
    out.push(
      clampVertex({ east_m: curr.east_m + sign * ox, north_m: curr.north_m + sign * oy }),
    );
  }
  return out;
}

export function offsetPolygonInward(vertices: PlanVertex[], distance: number): PlanVertex[] {
  return offsetPolygonOutward(vertices, -distance);
}

export function envelopeFillVolumeM3(areaTop: number, areaBottom: number, heightM: number): number {
  if (heightM <= 0) return 0;
  return (heightM / 3) * (areaTop + areaBottom + Math.sqrt(Math.max(0, areaTop * areaBottom)));
}

export function estimateEnvelopeFillM3(
  sketch: PlanShapeSketch,
  heightM: number,
  wrapWidthM: number,
): number | null {
  if (!Number.isFinite(heightM) || heightM <= 0 || wrapWidthM <= 0) return null;
  const vol = estimateEnvelopeBermRingVolumeM3(sketch, wrapWidthM);
  return vol > 0 ? vol : null;
}

/** Crest cap width TW between outer and inner crest (symmetric isosceles trapezoid). */
export function envelopeBermCrestCapWidthM(wrapWidthM: number): number {
  return wrapWidthM / 3;
}

/** 1:1 slope rise H = (W − TW) / 2 for both berm faces. */
export function envelopeBermSlopeHeightM(wrapWidthM: number): number {
  if (wrapWidthM <= 0) return 0;
  const tw = envelopeBermCrestCapWidthM(wrapWidthM);
  return (wrapWidthM - tw) / 2;
}

/** Trapezoid cross-section area of berm strip (per meter of edge). */
export function envelopeBermCrossSectionAreaM2(wrapWidthM: number): number {
  const h = envelopeBermSlopeHeightM(wrapWidthM);
  const tw = envelopeBermCrestCapWidthM(wrapWidthM);
  if (h <= 0 || wrapWidthM <= 0) return 0;
  return (h * (wrapWidthM + tw)) / 2;
}

/** Berm ring volume: perimeter × trapezoid cross-section (variant A). */
export function estimateEnvelopeBermRingVolumeM3(
  sketch: PlanShapeSketch,
  wrapWidthM: number,
): number {
  const verts = shapeVerticesForEnvelope(sketch);
  if (verts.length < 3 || wrapWidthM <= 0) return 0;
  return polygonPerimeterM(verts) * envelopeBermCrossSectionAreaM2(wrapWidthM);
}

export function planVerticesCentroid(vertices: PlanVertex[]): PlanVertex {
  if (vertices.length === 0) return { east_m: 0, north_m: 0 };
  return {
    east_m: vertices.reduce((sum, v) => sum + v.east_m, 0) / vertices.length,
    north_m: vertices.reduce((sum, v) => sum + v.north_m, 0) / vertices.length,
  };
}

export function envelopeOuterVertices(sketch: PlanShapeSketch, wrapWidthM: number): PlanVertex[] {
  const topVerts = shapeVerticesForEnvelope(sketch);
  if (topVerts.length < 3 || wrapWidthM <= 0) return topVerts;
  return offsetPolygonOutward(topVerts, wrapWidthM);
}

/** Outer edge of berm sole — coincides with pad top footprint. */
export function envelopeBermSoleOuterVertices(sketch: PlanShapeSketch): PlanVertex[] {
  return shapeVerticesForEnvelope(sketch);
}

/** Inner edge of berm sole — inset W from pad edge on pad top. */
export function envelopeBermSoleInnerVertices(
  sketch: PlanShapeSketch,
  wrapWidthM: number,
): PlanVertex[] {
  const outer = envelopeBermSoleOuterVertices(sketch);
  if (outer.length < 3 || wrapWidthM <= 0) return outer;
  return offsetPolygonInward(outer, wrapWidthM);
}

/** Outer crest line at inset H = (W − TW) / 2 from pad edge (1:1 from boundary). */
export function envelopeBermCrestOuterVertices(
  sketch: PlanShapeSketch,
  wrapWidthM: number,
): PlanVertex[] {
  const outer = envelopeBermSoleOuterVertices(sketch);
  if (outer.length < 3 || wrapWidthM <= 0) return outer;
  return offsetPolygonInward(outer, envelopeBermSlopeHeightM(wrapWidthM));
}

/** Inner crest line at inset (W + TW) / 2 — symmetric 1:1 from sole inner edge. */
export function envelopeBermCrestInnerVertices(
  sketch: PlanShapeSketch,
  wrapWidthM: number,
): PlanVertex[] {
  const outer = envelopeBermSoleOuterVertices(sketch);
  if (outer.length < 3 || wrapWidthM <= 0) return outer;
  const tw = envelopeBermCrestCapWidthM(wrapWidthM);
  return offsetPolygonInward(outer, (wrapWidthM + tw) / 2);
}

export function shapeVerticesForEnvelope(sketch: PlanShapeSketch): PlanVertex[] {
  if (isPlanPolygon(sketch)) return sketch.vertices;
  return localPlanCorners(sketch).map((c) => ({ east_m: c.east_m, north_m: c.north_m }));
}

File: decision-matrix/backend/pad-earthwork-vendor/src/pad_earthwork/envelope.py
"""Envelope wrap (truncated pyramid) volume helpers."""

from __future__ import annotations

import math

from pad_earthwork.schemas import EnvelopeWrap, PlanVertex
from pad_earthwork.volume_plan import polygon_area_m2


def _edge_outward_normal(dx: float, dy: float) -> tuple[float, float]:
    length = math.hypot(dx, dy)
    if length < 1e-12:
        return 0.0, 0.0
    return dy / length, -dx / length


def offset_polygon_outward(vertices: list[PlanVertex], distance: float) -> list[PlanVertex]:
    """Outward offset for CCW polygon vertices."""
    n = len(vertices)
    if n < 3 or distance <= 0:
        return [PlanVertex(east_m=v.east_m, north_m=v.north_m) for v in vertices]
    max_miter = distance * 4.0
    out: list[PlanVertex] = []
    for i in range(n):
        prev = vertices[(i - 1) % n]
        curr = vertices[i]
        nxt = vertices[(i + 1) % n]
        e1x = curr.east_m - prev.east_m
        e1y = curr.north_m - prev.north_m
        e2x = nxt.east_m - curr.east_m
        e2y = nxt.north_m - curr.north_m
        n1x, n1y = _edge_outward_normal(e1x, e1y)
        n2x, n2y = _edge_outward_normal(e2x, e2y)
        bx = n1x + n2x
        by = n1y + n2y
        bl = math.hypot(bx, by)
        if bl < 1e-9:
            ox, oy = n1x * distance, n1y * distance
        else:
            bx /= bl
            by /= bl
            dot = n1x * bx + n1y * by
            scale = distance / dot if abs(dot) > 1e-6 else distance
            if not math.isfinite(scale) or scale < 0 or scale > max_miter:
                ox, oy = n1x * distance, n1y * distance
            else:
                ox, oy = bx * scale, by * scale
        out.append(
            PlanVertex(
                east_m=max(-500.0, min(500.0, curr.east_m + ox)),
                north_m=max(-500.0, min(500.0, curr.north_m + oy)),
            )
        )
    return out


def envelope_fill_volume_m3(area_top: float, area_bottom: float, height_m: float) -> float:
    """Truncated pyramid: upper base = top contour, lower base = outward offset."""
    if height_m <= 0:
        return 0.0
    return (height_m / 3.0) * (
        area_top + area_bottom + math.sqrt(max(0.0, area_top * area_bottom))
    )


def _vertices_to_points(vertices: list[PlanVertex]) -> list[tuple[float, float]]:
    return [(v.east_m, v.north_m) for v in vertices]


def compute_envelope_volumes(
    top_vertices: list[PlanVertex],
    height_m: float,
    envelope: EnvelopeWrap,
) -> tuple[float, float, list[tuple[float, float]], list[str]]:
    """Returns fill_m3, footprint_area_m2 (bottom), outer corners local, warnings."""
    warnings: list[str] = ["envelope_volume_is_truncated_pyramid_approximation"]
    area_top = polygon_area_m2(_vertices_to_points(top_vertices))
    outer = offset_polygon_outward(top_vertices, envelope.wrap_width_m)
    area_bottom = polygon_area_m2(_vertices_to_points(outer))
    fill_m3 = envelope_fill_volume_m3(area_top, area_bottom, height_m)
    outer_local = [(v.east_m, v.north_m) for v in outer]
    return fill_m3, area_bottom, outer_local, warnings


def top_vertices_from_rectangle_sketch(sketch) -> list[PlanVertex]:
    from pad_earthwork.volume_plan import plan_corners_local_m

    return [
        PlanVertex(east_m=c[0], north_m=c[1]) for c in plan_corners_local_m(sketch)
    ]


def top_vertices_from_polygon_sketch(sketch) -> list[PlanVertex]:
    return list(sketch.vertices)


def outer_vertices_local_m(top_vertices: list[PlanVertex], wrap_width_m: float) -> list[tuple[float, float]]:
    outer = offset_polygon_outward(top_vertices, wrap_width_m)
    return [(v.east_m, v.north_m) for v in outer]
